### PR TITLE
Enable PowerShell placeholders

### PR DESCRIPTION
**Pull request by AnatoliB**

### CODE REVIEW

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: No need to update the list here. This is legacy before we had placeholder pools. I will remove this soon.

**Review by pragnagopa — DISMISSED** [review]
Added a minor comment.

**Comment by pragnagopa** [comment]
@fabiocav - Should we hold off merging new changes into dev?

**Review by mhoeger — DISMISSED** [review]
✅ lgtm, assuming pools work is also there for needed powershell versions! 

I also think this is already there but wanted to double check to be overly cautious: with pools matching, one version of powershell will never land on a placeholder with a different version of powershell? I know there are sometimes inexact matches. bringing this up because javascript is highly compatible between node versions and issues would be unlikely to surface

**Comment by pragnagopa** [comment]
> I also think this is already there but wanted to double check to be overly cautious: with pools matching, one version of powershell will never land on a placeholder with a different version of powershell? I know there are sometimes inexact matches. bringing this up because javascript is highly compatible between node versions and issues would be unlikely to surface

That is correct. If powershell version is not matched, it only falls back to host and does not use antoher version of poweshell. Same is true for node.

**Comment by AnatoliB** [comment]
> I know there are sometimes inexact matches

The notion of inexact matches sounds scary to me. We separated PS 6 and PS 7 for a reason. Even though there is a high degree of compatibility, they are still different, and I would rather take a cold start hit than a wrong PS version... I already implemented placeholder IDs for PowerShell in Antares, and I believe mixing PS versions is not supposed to happen. If it does happen, I will consider this an urgent bug, and we will disable PS placeholders ASAP.

File: src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs
Comment (AnatoliB) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: Great, this is what I suspected. I will revert this change.

Comment (AnatoliB) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: No, wait. @pragnagopa, reverting this breaks the `RpcInitializationService_Initializes_RpcServerAndChannels_Windows_PlaceHolderMode test`. The following expectation is not met:
``` csharp
_mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(RpcWorkerConstants.PowerShellLanguageWorkerName), Times.Once);
```
Do you think this is a correct expectation?

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/RpcInitializationService.cs]: Like rest of the languages in the test, except of Java this needs to 
`mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(RpcWorkerConstants.PowerShellLanguageWorkerName), Times.Never);`

**Comment by pragnagopa** [comment]
> The notion of inexact matches sounds scary to me.

@AnatoliB / @mhoeger - There is no inexact matches on the versions. I will share the link to matching logic in Antares offline.

@AnatoliB - Did you try these changes on private stamp for both PS6 and PS7?

**Comment by AnatoliB** [comment]
> @AnatoliB - Did you try these changes on private stamp for both PS6 and PS7?

In progress. I will not merge the PR until I'm done with testing.

**Comment by mhoeger** [comment]
Sorry for the confusion - thanks for clarifying @pragnagopa :)

**Comment by AnatoliB** [comment]
@pragnagopa I'm testing this on my private stamp, and I found a problem. When an exactly matching placeholder for PS6 or PS7 is available, the correct PS worker is used. However, when there is no exact placeholder available, the generic Functions v2 placeholder is used, and for some reason it starts a PS6 worker even though the site config specifies PS7. The most interesting thing is that `FUNCTIONS_WORKER_RUNTIME_VERSION` printed out of the invoked function contains `~7`, as expected, so it is actually injected properly. Any idea what is missing here?

My first guess is that perhaps the Functions host tries to determine which worker to run _before_ the `FUNCTIONS_WORKER_RUNTIME_VERSION` variable is injected. I will check the code, but does this sound plausible?

**Comment by AnatoliB** [comment]
> perhaps the Functions host tries to determine which worker to run _before_ the `FUNCTIONS_WORKER_RUNTIME_VERSION` variable is injected.

@pragnagopa, @mhoeger, @brettsam I've just heard that there is actually a known race condition in this area. Do you think this looks related? What is the status of this issue?

**Comment by pragnagopa** [comment]
>  When an exactly matching placeholder for PS6 or PS7 is available, the correct PS worker is used. However, when there is no exact placeholder available, the generic Functions v2 placeholder is used, 

Good that V2 host placeholder is used. This eliminates issue with matching logic

> and for some reason it starts a PS6 worker even though the site config specifies PS7. The most interesting thing is that FUNCTIONS_WORKER_RUNTIME_VERSION printed out of the invoked function contains ~7, as expected, so it is actually injected properly. Any idea what is missing here?

FUNCTIONS_WORKER_RUNTIME_VERSION is set to ~7 - indicates environment is loaded correctly. When using V2 host placeholder, powershell language worker process should start after specialization completes which is the flow today.  Adding more logging in [worker config parsing](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs) might help debug this further.

> I've just heard that there is actually a known race condition in this area. Do you think this looks related? What is the status of this issue?

I am not aware of open issues. Also, FUNCTIONS_WORKER_RUNTIME_VERSION is set to ~7 confirms that environment is loaded correctly. If we do identify a race condition, we should track an issue. Can you please link exiting issue here or open a new one if you have confirmed this is indeed a race?

**Comment by AnatoliB** [comment]
>  However, when there is no exact placeholder available, the generic Functions v2 placeholder is used,

Correction: I mentioned running into this on Functions V2, but it was in fact V3 (I rebased the same branch on V3, as V2 does not even have PS7).

Discussed this offline with Pragna, and #4574 seems to be the root cause. /cc @soninaren

**Comment by fabiocav** [comment]
This was originally targeting `dev`, which is now 3.0. Do you need to retarget? If so, please change the base branch and reopen the PR. Thanks!

**Comment by AnatoliB** [comment]
Thank you Fabio, this is in fact intentional. When I started this PR, I intended to target 2.0, but later discovered that there are special challenges in 3.0 that I want to be addressed first. Coincidentally, `dev` became 3.0, so these changes are on the right branch. After this one is merged, I will port this change to 2.0 as well.

**Review by pragnagopa — APPROVED** [review]
Please do let me know if you got a chance to test this change + latest changes in dev on a private stamp.

**Comment by AnatoliB** [comment]
> Please do let me know if you got a chance to test this change + latest changes in dev on a private stamp.

In progress, will finish later today.

**Comment by AnatoliB** [comment]
> Please do let me know if you got a chance to test this change + latest changes in dev on a private stamp.

@pragnagopa I'm observing the same issue when testing on my private stamp. Apparently, the #4574 fix was not enough:
```
2020-05-05 20:03:09.1273290  FUNCTIONS_WORKER_RUNTIME: powershell
2020-05-05 20:03:09.1273565  FUNCTIONS_WORKER_RUNTIME_VERSION: ~7
...
2020-05-05 20:03:09.2579491  Starting worker process (updated):D:\Program Files (x86)\dotnet\dotnet.exe  "D:\Program Files (x86)\SiteExtensions\Functions\3.0.13353\workers\powershell\6/Microsoft.Azure.Functions.PowerShellWorker.dll" ...
```
I'm running the code from this branch (already rebased on top of your fix). I will email the query.

**Comment by AnatoliB** [comment]
Tested on a private stamp, looks good. The only problem I found is resolved by https://github.com/Azure/azure-functions-host/pull/5990.

**Pull request merged**
